File: src/main/scala/cracking/BubbleSort.scala
package cracking

import scala.util.control._
import util.control.Breaks._

object BubbleSort {

  def main(args: Array[String]): Unit = {
   // sort(Array(3, 2, 1))
   // sort(Array(1, 2, 3))
    sort(Array(4, 2, 3, 1, 5))
  }

  def sort(arrays: Array[Int]) = {
    var swaps = 0
    for (i <- 0 until arrays.length) {
      var numberOfSwaps = 0
      val loop = new Breaks
      breakable {
        for(j <- 0 until arrays.length -1) {
          if(arrays(j) > arrays(j+1)) {
            swap(arrays, j, j+1)
            numberOfSwaps += 1
            swaps += 1
          }
          println(arrays.mkString(","))
          //swaps += numberOfSwaps
          if(numberOfSwaps == 0) {
            println("break")
            break
          }
        }
      }
    }

    println(s"Array is sorted in $swaps swaps.")
    println(s"First Element: ${arrays(0)}")
    println(s"Last Element: ${arrays(arrays.length - 1)}")
  }

  def swap(arrays: Array[Int], i: Int, j:Int) = {
    val temp = arrays(i)
    arrays(i) = arrays(j)
    arrays(j) = temp
  }
}

File: src/main/java/sort/BubbleSort.java
package sort;

public class BubbleSort {

    public static void main(String []args) {
        sort(new int[]{4, 2, 3, 1, 5});
        sort(new int[]{1,2,3});  
    }

    public static void sort(int []a) {
        int n = a.length;
        int swaps = 0;
        for (int i = 0; i < n; i++) {
            // Track number of elements swapped during a single array traversal
            int numberOfSwaps = 0;

            for (int j = 0; j < n - 1; j++) {
                // Swap adjacent elements if they are in decreasing order
                if (a[j] > a[j + 1]) {
                    swap(a, j, j + 1);
                    numberOfSwaps++;
                    swaps++;
                }
            }

            // If no elements were swapped during a traversal, array is sorted
            if (numberOfSwaps == 0) {
                break;
            }
        }

        System.out.println("Array is sorted in " + swaps + " swaps.");
        System.out.println("First Element: " + a[0]);
        System.out.println("Last Element: " + a[a.length - 1]);
    }
            
    private static void swap(int []a, int i, int j) {
        int temp = a[i];
        a[i] = a[j];
        a[j] = temp;
    }
}
